Modular trashbin api

icewind1991 opened the pull request:
Modular api for trashbin backends

- backends are registered in info.xml
- backends are selected based on storage class
- existing trash logic has been wrapped in a "legacy" backend
- trash dav api has been adjusted to the new api

- [ ] schema for info.xml will have to be updated on the appstore

icewind1991 committed:
fix select statement
fix select statement
Make trashbin api modules

Apps can register trashbin backends for specific storages,
allowing them to modify trashbin behavior for storages

The old trashbin implementation has been wrapped in a "legacy" backend,
for future work this can be replaced with a new backend that better handles
shares while still keeping the legacy backend around to keep existing trash
from being accessible

Signed-off-by: Robin Appelman <robin@icewind.nl>

icewind1991 committed:
fix trashbin previews for modular api

Signed-off-by: Robin Appelman <robin@icewind.nl>

icewind1991 committed:
use name in fileinfo if set

Signed-off-by: Robin Appelman <robin@icewind.nl>

icewind1991 committed:
expose storagefactory in server container

Signed-off-by: Robin Appelman <robin@icewind.nl>

icewind1991 committed:
pass `Cache::getMoveInfo` along cache wrappers

fixes cross storage move in some cases

Signed-off-by: Robin Appelman <robin@icewind.nl>

icewind1991 committed:
fix error message for folder preview request

Signed-off-by: Robin Appelman <robin@icewind.nl>

icewind1991 committed:
fix preview controller tests

Signed-off-by: Robin Appelman <robin@icewind.nl>

icewind1991 committed:
fix trash storage wrapper tests

Signed-off-by: Robin Appelman <robin@icewind.nl>

icewind1991 commented:
@rullzer  @MorrisJobke  please review

rullzer commented on apps/files_trashbin/lib/Trash/TrashManager.php:
```diff
@@ -0,0 +1,125 @@
+	private $trashPaused = false;
+
+	public function registerBackend(string $storageType, ITrashBackend $backend) {
+		$this->backends[$storageType] = $backend;
```
Shoudl we check if the storagetype is actually valid?

icewind1991 replied:
I don't see to much of an advantage of adding such a check

rullzer reviewed (commented):
Gave it a quick look. will test tomorrow

rullzer commented on apps/files_trashbin/lib/Trash/ITrashBackend.php:
```diff
@@ -0,0 +1,86 @@
+	 * @return ITrashItem[]
+	 * @since 15.0.0
+	 */
+	public function listTrashFolder(IUser $user, ITrashItem $folder): array;
```
Is restore item doesn't need the user. Then why does this need the user?

rullzer commented on apps/files_trashbin/lib/Trash/TrashItem.php:
```diff
@@ -0,0 +1,187 @@
+	}
+
+	public function getOriginalLocation(IUser $user): string {
+		return $this->orignalLocation;
```
I'm assume the `$user` here is to accomodate for shared trash folders?

icewind1991 replied:
originally yes, but since the `$user` is already known when creating the trash item I've removed it from here.

rullzer commented on apps/files_trashbin/lib/Trash/TrashManager.php:
```diff
@@ -0,0 +1,125 @@
+
+	public function listTrashRoot(IUser $user): array {
+		$items = array_reduce($this->getBackends(), function (array $items, ITrashBackend $backend) use ($user) {
+			return array_merge($items, $backend->listTrashRoot($user));
```
Maybe for a bit of hardening loop over all the items and check if they are of the proper interface?

rullzer commented:
I like it!

I took it for a test and it works. (Also old stuff is found correctly).
Cool stuff.

icewind1991 commented on apps/files_trashbin/lib/AppInfo/Application.php:
```diff
@@ -61,5 +64,36 @@ public function __construct (array $urlParams = []) {
+
+					try {
+						$backendObject = $server->query($class);
+						$trashManager->registerBackend($for, $backendObject);
```
An invalid type will already be catched and logged by the try/catch

icewind1991 commented on lib/public/IServerContainer.php:
```diff
@@ -584,4 +584,10 @@ public function getRemoteApiFactory();
+	 * @return \OCP\Files\Storage\IStorageFactory
+	 * @since 15.0.0
+	 */
+	public function getStorageFactory();
```
Since any app registering their own mounts have a use for this I would prefer to have this "properly" in the public api

rullzer reviewed (approved):
Lets do this

icewind1991 committed:
remove user parameter from `ITrashItem::getOriginalLocation`

since the target user is known at the time of creating the trashitem
the original location can already be adjusted on a per user level

Signed-off-by: Robin Appelman <robin@icewind.nl>

icewind1991 committed:
store user for trashitem in the trashitem

Signed-off-by: Robin Appelman <robin@icewind.nl>

icewind1991 commented:
changed the interface around a bit to store the user in the trashitem, this removes the need to pass the user to a method in most cases while still having it around for backends that need it

icewind1991 commented:
@MorrisJobke @schiessle please review

icewind1991 commented:
ping @MorrisJobke @schiessle

MorrisJobke reviewed (approved):
Tested and works 👍
